### PR TITLE
Update with Marco's comments #6

### PR DESCRIPTION
**Pull request by mrdeep1**

### CODE REVIEW

File: draft-ietf-core-new-block.xml [hunk 1]
```diff
@@ -1742,6 +1762,8 @@ o  Reference: [RFCXXXX]]]></artwork>
 
       <?rfc include='reference.RFC.8949'?>
 
+      <?rfc include='reference.RFC.8974'?>
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml:1779]: Please move this reference to the Informative References.

File: draft-ietf-core-new-block.xml [hunk 2]
```diff
@@ -1529,47 +1552,42 @@ X<--: Message loss (response)
 
           <t>The client realizes that not all the payloads of the response
           have been returned. The client then asks for the missing blocks in
-          one go (<xref target="b12non3"></xref>). However, because the
-          original request is spread over multiple payloads using Q-Block1,
-          the entire FETCH body has to be used to make the request for the
-          missing payloads.</t>
+          one go (<xref target="b12non3"></xref>). Note that, following Section
+          2.4 of <xref target="RFC7959"></xref>, the FETCH request does not
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]: Which part of 2.4 are you referring to? Why  not just saying that we are mimic'ing the behavior in Section 3.3?

File: draft-ietf-core-new-block.xml [hunk 3]
```diff
@@ -1228,23 +1214,64 @@ X<--: Message loss (response)
...
+         |     [[NON_MAX_RETRASMIT exceeded.  Server stops requesting for
+         |       missing blocks and releases partial body]]
+         |   ...    |
+]]></artwork>
+            </figure></t>
         </section>
       </section>
 
       <section title="Q-Block2 Option">
+        <t>These examples include the Observe Option to demonstrate how the
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <t>These examples include the Observe Option to demonstrate how that option
```

File: draft-ietf-core-new-block.xml [hunk 4]
```diff
@@ -1228,23 +1214,64 @@ X<--: Message loss (response)
...
+        Observe Option is included and used. The Observe Option is not
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        is used. The Observe Option is not
```

File: draft-ietf-core-new-block.xml [hunk 5]
```diff
@@ -1228,23 +1214,64 @@ X<--: Message loss (response)
...
-          implementation specific.</t>
+        </section>
+
+        <section title="Handling Recovery with Failure">
+          <t><xref target="B3non3"></xref> depicts an example of a NON PUT
+          request conveying Q-Block1 Option where recovery takes place, but
+          eventually fails.</t>
+
+          <t><figure anchor="B3non3"
+              title="Example of NON Request with Q-Block1 Option (With eventual failure)">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
              title="Example of NON Request with Q-Block1 Option (With Eventual Failure)">
```

File: draft-ietf-core-new-block.xml [hunk 6]
```diff
@@ -1208,11 +1193,12 @@ X<--: Message loss (response)
           <t>On seeing a payload from the next set of payloads, the server
           realizes that some blocks are missing from the previous MAX_PAYLOADS
           payloads and asks for the missing blocks in one go (<xref
-          target="B3non2"></xref>). It does so by indicating which blocks have
-          not been received in the data portion of the response. The token
-          used in the response should be the token that was used in the last
-          block number received payload. The client can then derive the
-          Request-Tag by matching the token with the sent request.</t>
+          target="B3non2"></xref>). It does so by indicating which blocks from
+          the previous MAX_PAYLOADS have not been received in the data portion
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
          the previous MAX_PAYLOADS payloads have not been received in the data portion
```

File: draft-ietf-core-new-block.xml [hunk 7]
```diff
@@ -859,9 +837,14 @@ missing-block-number = uint
     </section>
 
     <section anchor="cc" title="Congestion Control">
-      <t>The transmission of the payloads of a body either SHOULD all be
-      Confirmable or all be Non-confirmable. This is meant to simplify the
-      congestion control procedure.</t>
+      <t>The transmission of the payloads of a body over UDP or DTLS connection
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
      <t>The transmission of the payloads of a body over an unreliable transport      
```

File: draft-ietf-core-new-block.xml [hunk 8]
```diff
@@ -859,9 +837,14 @@ missing-block-number = uint
     </section>
 
     <section anchor="cc" title="Congestion Control">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
    <section anchor="cc" title="Congestion Control for Unreliable Transports">
```

File: draft-ietf-core-new-block.xml [hunk 9]
```diff
@@ -859,9 +837,14 @@ missing-block-number = uint
...
-      <t>The transmission of the payloads of a body either SHOULD all be
-      Confirmable or all be Non-confirmable. This is meant to simplify the
-      congestion control procedure.</t>
+      <t>The transmission of the payloads of a body over UDP or DTLS connection
+      SHOULD either all be Confirmable or all be Non-confirmable.
+      This is meant to simplify the congestion control procedure.</t>
+
+      <section anchor="cc-tcp" title="Reliable Transports">
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
```

File: draft-ietf-core-new-block.xml [hunk 10]
```diff
@@ -859,9 +837,14 @@ missing-block-number = uint
...
+        <t>There is no need for additional Congestion Control for reliable
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        <t>As a reminder, there is no need for CoAP-specific congestion control for reliable
```

File: draft-ietf-core-new-block.xml [hunk 11]
```diff
@@ -901,15 +885,18 @@ missing-block-number = uint
         target="RFC7252"></xref>).</t>
 
         <t>NON_RECEIVE_TIMEOUT is the maximum time to wait for a missing
-        payload before requesting retransmission. NON_RECEIVE_TIMEOUT has a
-        value of twice NON_TIMEOUT.</t>
+        payload before requesting retransmission. This is the initial timeout
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        payload before requesting retransmission. Every time the missing
        payload is re-requested, the timeout value doubles. The NON_RECEIVE_TIMEOUT
        has a default initial value of twice NON_TIMEOUT. NON_RECEIVE_TIMEOUT MUST
        always be greater than NON_TIMEOUT by at least one second.</t>
```

File: draft-ietf-core-new-block.xml [hunk 12]
```diff
@@ -901,15 +885,18 @@ missing-block-number = uint
...
+        for the first request of a missing payload. Every time the missing
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion

```

File: draft-ietf-core-new-block.xml [hunk 13]
```diff
@@ -901,15 +885,18 @@ missing-block-number = uint
...
+        payload is re-requested, the timeout value doubles. NON_RECEIVE_TIMEOUT
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion

```

File: draft-ietf-core-new-block.xml [hunk 14]
```diff
@@ -901,15 +885,18 @@ missing-block-number = uint
...
+        has a default value of twice NON_TIMEOUT. NON_RECEIVE_TIMEOUT MUST
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion

```

File: draft-ietf-core-new-block.xml [hunk 15]
```diff
@@ -901,15 +885,18 @@ missing-block-number = uint
...
+        always be greater than NON_TIMEOUT by at least one second.</t>
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion

```

Comment (boucadair) [reply draft-ietf-core-new-block.xml]: I wonder whether some text can be added to explain the rationale for the one second mentioned in the text.

File: draft-ietf-core-new-block.xml [hunk 16]
```diff
@@ -901,15 +885,18 @@ missing-block-number = uint
...
 
         <t>NON_MAX_RETRANSMIT is the maximum number of times a request for the
         retransmission of missing payloads can occur without a response from
-        the remote peer. After this occurs, the peer SHOULD consider the body
-        stale and remove all references to it. By default, NON_MAX_RETRANSMIT
-        has the same value as MAX_RETRANSMIT (Section 4.8 of <xref
-        target="RFC7252"></xref>).</t>
+        the remote peer. After this occurs, the local peer SHOULD consider the
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        the remote peer. After this occurs, the local endpoint SHOULD consider the
```

File: draft-ietf-core-new-block.xml [hunk 17]
```diff
@@ -850,7 +826,9 @@ missing-block-number = uint
           <t hangText="Implementation Note:">To minimize on the number of
           tokens that have to be tracked by clients, it is recommended that
           the bottom 32 bits is kept the same for the same body and the upper
-          32 bits contains the individual payload number. <vspace
+          32 bits contains the individual payload number. This allows the
+          client to be alleviated from keeping all the per-request-state as in
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
          client to be alleviated from keeping all the per-request-state, e.g., in 
```

Make it clear this is just an example.

File: draft-ietf-core-new-block.xml [hunk 18]
```diff
@@ -541,13 +528,12 @@
...
+            and Q-Block1 to specify the block NUM, SZX, and M bit as
             appropriate.</t>
 
-            <t>The Request-Tag value to use is determined from the token in
-            the 4.08 (Request Entity Incomplete) response and then finding the
-            matching client request which contains the Request-Tag that is
-            being used for this Q-Block1 body.</t>
+            <t>The Request-Tag value to use is determined by taking the token
+            in the 4.08 (Request Entity Incomplete) response, locating the
+            matching client request and then using its Request-Tag.</t>
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
            matching client request, and then using its Request-Tag.</t>
```

File: draft-ietf-core-new-block.xml [hunk 19]
```diff
@@ -412,20 +406,8 @@
...
-        could be given in the option, i.e., it indicates that the current
-        block is the first and only block of the transfer (block number is set
-        to 0, M is unset). However, in contrast to the explicit value 0, which
-        would indicate a size of the block (SZX) of 0, and thus a size value
-        of 16 bytes, there is no specific explicit size implied by the absence
-        of the option -- the size is left unspecified. (As for any uint, the
-        explicit value 0 is efficiently indicated by a zero-length option;
-        this, therefore, is different in semantics from the absence of the
-        option).</t>
+        target="RFC7959"></xref> with Q-Block1 substituted for Block1 and
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]: I suggest to maintain the old wording of this section and limit the chnages to those required to address the WGLC comments. 

As a reminder, we included some tetx from 7959 on purpose. This is why we have the following: 

"Some text from [RFC7959] is reused for readers convenience."

File: draft-ietf-core-new-block.xml [hunk 20]
```diff
@@ -455,9 +437,13 @@
 
         <t>For Confirmable transmission, the server continues to acknowledge
         each packet, but a response is not required (whether separate or
-        piggybacked) until successful receipt of the body or, if some of the
-        payloads are sent as Non-confirmable and have not arrived, a
-        retransmit missing payloads response is needed.</t>
+        piggybacked) until successful receipt of the body by the server. For
+        Non-confirmable transmission no response is required until the
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        Non-confirmable transmission, no response is required until the
```

File: draft-ietf-core-new-block.xml [hunk 21]
```diff
@@ -455,9 +437,13 @@
...
+        successful receipt of the body by the server or some of the payloads
+        have not arrived after a timeout and a retransmit missing payloads
+        response is needed. For reliable connections <xref
+        target="RFC8323"></xref> a response is not required until successful
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        target="RFC8323"></xref>), a response is not required until successful
```

File: draft-ietf-core-new-block.xml [hunk 22]
```diff
@@ -455,9 +437,13 @@
...
         <t>For Confirmable transmission, the server continues to acknowledge
         each packet, but a response is not required (whether separate or
-        piggybacked) until successful receipt of the body or, if some of the
-        payloads are sent as Non-confirmable and have not arrived, a
-        retransmit missing payloads response is needed.</t>
+        piggybacked) until successful receipt of the body by the server. For
+        Non-confirmable transmission no response is required until the
+        successful receipt of the body by the server or some of the payloads
+        have not arrived after a timeout and a retransmit missing payloads
+        response is needed. For reliable connections <xref
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        response is needed. For reliable transports (e.g., <xref
```

File: draft-ietf-core-new-block.xml [hunk 23]
```diff
@@ -668,13 +656,17 @@
 
         <t>A client SHOULD only maintain a partial body (missing payloads) for
         up to NON_PARTIAL_TIMEOUT (<xref target="cc-non"></xref>) or as
-        defined by the Max-Age Option (or its default), whichever is the
-        less.</t>
+        defined by the Max-Age Option (or its default of 60 seconds), whichever
```
Comment (boucadair) [inline draft-ietf-core-new-block.xml]:
```suggestion
        defined by the Max-Age Option (or its default of 60 seconds (Section 5.6.1 of <xref target="RFC7252"></xref>)), whichever
```

**Review by boucadair — CHANGES_REQUESTED** [review]
Thanks for the proposal. Please consider these comments. Once done, will merge.

File: draft-ietf-core-new-block.xml (hunk 2 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Oops - I meant 2.7 - fixed

File: draft-ietf-core-new-block.xml (hunk 7 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: I thought about using unreliable, but then CoAP supports reliable transmission when using CON.  Hence my text, as you could argue CoAP is a reliable transport when using UDP.

File: draft-ietf-core-new-block.xml (hunk 8 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Same comment about the usage of Unreliable as you can argue CoAP is a reliable transport when using UDP when using CON.

File: draft-ietf-core-new-block.xml (hunk 9 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Have now removed the section wrapper after doing the following change.

File: draft-ietf-core-new-block.xml (hunk 19 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Text reverted

File: draft-ietf-core-new-block.xml (hunk 15 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: Text added

File: draft-ietf-core-new-block.xml (hunk 19 above)
Comment (boucadair) [reply draft-ietf-core-new-block.xml]: Thanks

File: draft-ietf-core-new-block.xml (hunk 7 above)
Comment (boucadair) [reply draft-ietf-core-new-block.xml]: CoAP is not a transport.

File: draft-ietf-core-new-block.xml (hunk 8 above)
Comment (boucadair) [reply draft-ietf-core-new-block.xml]: Same as above. We do have reliable and unreliable transport for CoAP. I don't think there is a confusion out there.

File: draft-ietf-core-new-block.xml (hunk 7 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: OK - Change included.

File: draft-ietf-core-new-block.xml (hunk 8 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: OK - Text changed.

File: draft-ietf-core-new-block.xml (hunk 11 above)
Comment (mrdeep1) [reply draft-ietf-core-new-block.xml]: This chunk of text has been re-worded.

**Comment by mrdeep1** [comment]
Updated text pushed.

**Comment by boucadair** [comment]
Thanks.

**Pull request merged**